src/main/java/uk/gov/hmcts/reform/hmc/service/AccessControlServiceImpl.java

```diff
@@ -0,0 +1,169 @@
+    }
+
+    private List<RoleAssignment> verifyRoleAccess(List<String> requiredRoles, Long hearingId) {
+        RoleAssignments roleAssignments = roleAssignmentService.getRoleAssignments(securityUtils.getUserId());
+        if (roleAssignments.getRoleAssignments().isEmpty()) {
+            throw new ResourceNotFoundException(String.format(ROLE_ASSIGNMENTS_NOT_FOUND, securityUtils.getUserId()));
+        }
+        List<RoleAssignment> filteredRoleAssignments = filterRoleAssignments(roleAssignments);
+        if (filteredRoleAssignments.isEmpty()) {
+            throw new InvalidRoleAssignmentException(ROLE_ASSIGNMENT_INVALID_ROLE);
+        }
+
+        verifyRequiredRolesExists(requiredRoles, filteredRoleAssignments);
+        return filteredRoleAssignments;
+    }
+
+    @Override
+    public void verifyHearingCaseAccess(Long hearingId, List<String> requiredRoles) {
+        CaseHearingRequestEntity caseHearingRequestEntity = caseHearingRequestRepository.getCaseHearing(hearingId);
+        if (caseHearingRequestEntity != null) {
+            verifyCaseAccess(caseHearingRequestEntity.getCaseReference(), requiredRoles, hearingId);
+        }
+    }
+
+    private void verifyRequiredRolesExists(List<String> requiredRoles, List<RoleAssignment> filteredRoleAssignments) {
+        if (!requiredRoles.isEmpty()) {
+            boolean containsRequiredRoles = filteredRoleAssignments.stream()
+                .anyMatch(roleAssignment -> requiredRoles.contains(roleAssignment.getRoleName()));
+
+            if (!containsRequiredRoles) {
+                throw new InvalidRoleAssignmentException(ROLE_ASSIGNMENT_MISSING_REQUIRED);
+            }
+        }
+    }
+
+    private void verifyHearingStatus(List<RoleAssignment> filteredRoleAssignments,
+                                     String caseTypeId,
+                                     Long hearingId) {
+        if (hearingId != null) {
+            String status = hearingRepository.getStatus(hearingId);
+            if (filteredRoleAssignments
+                .stream()
+                .filter(roleAssignment -> verifyCaseType(roleAssignment, caseTypeId))
+                .allMatch(roleAssignment -> LISTED_HEARING_VIEWER.equals(roleAssignment.getRoleName()))
+                && !HearingStatus.LISTED.name().equals(status)) {
+                throw new InvalidRoleAssignmentException(ROLE_ASSIGNMENT_INVALID_STATUS);
+            }
+        }
+    }
+
+    @SuppressWarnings("java:S2789")
+    private boolean verifyCaseType(RoleAssignment roleAssignment, String caseTypeId) {
+        Optional<String> roleCaseType = roleAssignment.getAttributes().getCaseType();
+        if (roleCaseType == null
+            || roleCaseType.isEmpty()
+            || caseTypeId == null) {
+            return true;
+        }
+        return roleCaseType.orElse("").equals(caseTypeId);
+    }
+
+    private List<RoleAssignment> filterRoleAssignments(RoleAssignments roleAssignments) {
```

Comment:
Ah missed this on previous review - we wanted to add the check that the role assignments are valid for the current time, like we do in Data Store:
https://github.com/hmcts/ccd-data-store-api/blob/e6d5579f206077c006f9ca7999ffbecca9bc89f9/src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/RoleAssignment.java#L39-L43